src/framework/mpe/events.h

```diff
@@ -0,0 +1,171 @@
+    {
+        m_arrangementCtx.nominalDuration = nominalDuration;
+        m_arrangementCtx.nominalTimestamp = nominalTimestamp;
+        m_arrangementCtx.voiceLayerIndex = voiceIdx;
+
+        m_pitchCtx.pitchClass = pitchClass;
+        m_pitchCtx.octave = octave;
+
+        m_expressionCtx.articulations = articulationsApplied;
+        m_expressionCtx.dynamic = dynamicType;
+
+        ArrangementPatternList arrangementPatterns = m_expressionCtx.articulations.arrangementPatterns();
+        calculateActualDuration(arrangementPatterns);
+        calculateActualTimestamp(arrangementPatterns);
+
+        calculatePitchCurve(standardPatterns.pitchPattern);
+
+        calculateExpressionCurve(standardPatterns.expressionPattern);
+    }
+
+    const ArrangementContext& arrangementCtx() const
+    {
+        return m_arrangementCtx;
+    }
+
+    const PitchContext& pitchCtx() const
+    {
+        return m_pitchCtx;
+    }
+
+    const ExpressionContext& expressionCtx() const
+    {
+        return m_expressionCtx;
+    }
+
+private:
+    void calculateActualTimestamp(const ArrangementPatternList& patterns)
+    {
+        m_arrangementCtx.actualTimestamp = m_arrangementCtx.nominalTimestamp;
+
+        if (patterns.empty()) {
+            return;
+        }
+
+        m_arrangementCtx.actualTimestamp -= ArrangementPattern::averageTimestampOffset(patterns);
+    }
+
+    void calculateActualDuration(const ArrangementPatternList& patterns)
+    {
+        m_arrangementCtx.actualDuration = m_arrangementCtx.nominalDuration;
+
+        if (patterns.empty()) {
+            return;
+        }
+
+        m_arrangementCtx.actualDuration *= ArrangementPattern::averageDurationFactor(patterns);
+    }
+
+    void calculatePitchCurve(const PitchPattern& standardPattern)
+    {
+        PitchPatternList appliedPatterns = m_expressionCtx.articulations.pitchPatterns();
+        PitchPattern::PitchOffsetMap appliedOffsetMap = PitchPattern::averagePitchOffsetMap(std::move(appliedPatterns));
+
+        for (const auto& pair : standardPattern.pitchOffsetMap) {
+            m_pitchCtx.pitchCurve.emplace(pair.first, pair.second + appliedOffsetMap[pair.first]);
+        }
+    }
+
+    void calculateExpressionCurve(const ExpressionPattern& standardPattern)
+    {
+        ExpressionPatternList appliedPatterns = m_expressionCtx.articulations.expressionPatterns();
+        ExpressionPattern::DynamicOffsetMap appliedOffsetMap = ExpressionPattern::averageDynamicOffsetMap(std::move(appliedPatterns));
+
+        float ratio = dynamicLevelFromType(m_expressionCtx.dynamic) / standardPattern.maxAmplitudeLevel;
+
+        for (const auto& pair : standardPattern.dynamicOffsetMap) {
+            m_expressionCtx.expressionCurve.emplace(pair.first, (pair.second * ratio) + appliedOffsetMap[pair.first]);
+        }
+    }
+
+    ArrangementContext m_arrangementCtx;
+    PitchContext m_pitchCtx;
+    ExpressionContext m_expressionCtx;
+};
+
+struct RestEvent
```

Comment:
If `NoteEvent` and `RestEvent` do not have a base class, how are `NoteEvent` and `RestEvent` to be stored and transferred? Not on the same list one after the other? On two different lists?

Reply:
It's going to be a single list of std::variant<NoteEvent, RestEvent>